docs/InstrumentationVerification/InstrumentationVerification.md

```diff
@@ -0,0 +1,134 @@
+
+3. As part of development of new bytecode instrumentation techniques, enabling breakpoint debugging of instrumented methods via [dnSpy](https://github.com/dnSpy/dnSpy).  
+
+
+# Usage
+
+To run a verification, simply:
+
+
+
+1. Add the environment variable `DD_TRACE_WRITE_INSTRUMENTATION_TO_DISK`, set to “true”.
+2. Run or restart the application we are instrumenting. Observe that as we instrument the app, more and more files will appear under [Datadog Logs Folder]/InstrumentationVerification/{ProcessName}_{ProcessID}_{ProcessCreationTime}
+3. Via command line, run the command: 
+
+    Datadog.InstrumentedAssemblyVerification.Standalone.exe [Full path of the folder generated in step 2] 
+
+
+    Once the command has finished executing, you will find new files under the InstrumentationVerification folder:
+
+    1. Logs file of the verification tool for each assembly will appear under a folder named **OUPUT_InstrumentedAssemblies/InstrumentationVerificationLogger_[DATE]_.log**)
+    2. The instrumented assemblies will appear under a folder named **OUPUT_InstrumentedAssemblies**
+
+**Note:** When handling a support case, the first two steps must be performed on the customer’s site. However, if the customer is willing to upload and send us the instrumentation output folder (which is conveniently placed inside the Datadog Logs folder), then the 3rd step can be performed offline.
+
+
+# Approach & High Level Design
+
+
+![alt_text](InstrumentationVerificationDesign.png "instrumentation verification high level design")
+
+
+The Instrumentation Verification process is divided into three main steps:
+
+
+
+1. At runtime, we write the following information to disk:
+    1. A copy of every assembly that was loaded 
+    2. Several “metadata” files which contains all the metadata changes that were performed via _IMetadataEmit2 / IMetadataAssemblyEmit._ The file format is textual, in order to assist with easy troubleshooting.
+    3. A “bytecode” file, which contains all the IL bytecode that was modified via _ICorProfilerInfo::SetILFunctionBody_ or _ICorProfilerFunctionControl::SetILFunctionBody,_ in binary format.
+
+    
+
+    The files in section (a) are written to a folder named **INPUT_OriginalAssemblies** whereas the files in sections (b) and (c) are written to a folder named **INPUT_InstrumentationLog.**
+
+
+    The logic of writing this information to disk resides within the Native Loader and is abstracted away inside proxy classes that implement the raw CLR Profiling API interfaces. As a result, the Tracer, Continuous Profiler, or any other client library that the Native Loader may dispatch calls to in the future does not need to know that IL Verification is taking place.
+
+
+    
+![alt_text](InstrumentationVerificationNativeWrappers.png "Instrumentation verification native wrappers")
```

Comment:
nit- on a dark background the arrows aren't super visible

Reply:
Fixed in 4c7c44dcf9bc4f5affbde6c20b8fd17d1735d225